### ISSUE TITLE
Build for WindowsXP is posible?

### ISSUE THREAD

**Issue by devisbal** [basiliscos/syncspirit#3]
Hi:

It will be possible to build for 32bits Windows XP this program syncspirit?
We do have been using SyncThing on Windows XP, but those just are the latest posible we could compile/build that was: V1.6.1 

BR

**Comment by basiliscos** [comment]
Hi,

I've made a i686-build, i.e. win32 (not win64) binary.  I'm not sure , that it works on `windows xp`, please try the attached files.

Let me know the outcome!

[windows-i686-support-libraries.zip](https://github.com/basiliscos/syncspirit/files/15047855/windows-i686-support-libraries.zip)

[syncspirit-daemon.zip](https://github.com/basiliscos/syncspirit/files/15047861/syncspirit-daemon.zip)

**Comment by basiliscos** [comment]
What I have found that my compiler (mingw) is most likely compiled with default target to windows vista. It is something like that:

https://github.com/msys2/MINGW-packages/blob/4a7eb3e90bc008e80274b7d2968269a5ea38156c/mingw-w64-headers-git/PKGBUILD#L58

So, if you can build mingw on yourself (on linux machine) it should probably work, because I was able to compile it.

Here are a few patches:

```
diff --git a/src/utils/network_interface.cpp b/src/utils/network_interface.cpp
index 01558d8..034168d 100644
--- a/src/utils/network_interface.cpp
+++ b/src/utils/network_interface.cpp
@@ -115,8 +115,7 @@ static uri_container_t _local_interfaces(logger_t &log, std::uint16_t port) noex
                 // IPv4
                 SOCKADDR_IN *ipv4 = reinterpret_cast<SOCKADDR_IN *>(address->Address.lpSockaddr);
 
-                char host[INET_ADDRSTRLEN] = {0};
-                inet_ntop(AF_INET, &(ipv4->sin_addr), host, INET_ADDRSTRLEN);
+                auto host = inet_ntoa(ipv4->sin_addr);
                 auto full = fmt::format("tcp://{}:{}/", host, port);
                 auto uri = utils::parse(full).value();
                 r.emplace_back(std::move(uri));
```

 ```
 --- a/src/osal.c
+++ b/src/osal.c
@@ -1090,7 +1090,7 @@ MDBX_INTERNAL_FUNC void osal_ioring_reset(osal_ioring_t *ior) {
     for (ior_item_t *item = ior->pool; item <= ior->last;) {
       if (!HasOverlappedIoCompleted(&item->ov)) {
         assert(ior->overlapped_fd);
-        CancelIoEx(ior->overlapped_fd, &item->ov);
+        CancelIo(ior->overlapped_fd);
       }
       if (item->ov.hEvent && item->ov.hEvent != ior)
         ior_put_event(ior, item->ov.hEvent);
```

+ a conan profile is attached 
[xp.txt](https://github.com/basiliscos/syncspirit/files/15049738/xp.txt)


More building instructions u can find in `doc/building.md`. 

PS. You can find whether it need windows-vista via the command `grep -a GetTickCount64 *.dll`

**Comment by basiliscos** [comment]
will try to build the mingw compiler by myself, not sure that it will be successful :)

**Comment by basiliscos** [comment]
```
-> % find . -iname 'osal.c'
./lib/mbdx/src/osal.c
```

**Comment by basiliscos** [comment]
I think I was able to build for windowx xp, but unfortunately `libmbdx` reports some error during startup. I have asked the libmbdx author whether it is possible to add windows xp support. Will let u know 

[syncspirit-daemon.exe.zip](https://github.com/basiliscos/syncspirit/files/15058740/syncspirit-daemon.exe.zip)

**Comment by basiliscos** [comment]
I'll try to use a different db-engine (switcheable) for winxp, but probably in a next version. 

The engines under consideration are:

https://github.com/pmwkaa/ioarena/tree/master/db

**Comment by basiliscos** [comment]
[syncspirit.zip](https://github.com/basiliscos/syncspirit/files/15065447/syncspirit.zip)

I think I've managed to do a proper build, please try @devisbal . 

If it works, I'll do more official release `v0.3.1` a bit later + instructions how to build.

**Comment by basiliscos** [comment]
>  We do use defined direct connection to the SyncThing Server. [IP:port] We do not have a Discover Server either.

This is correct `sync-spirit` behavior, as the "direct discovery" feature  not implemented , please file a separate issue. Only relay/global/local discoveries work.

However, you can take the device id and ip address from `sync-spirit` and copy it to other client... and it will connect to syncthing and the sync should start.

**Comment by basiliscos** [comment]
> I would like to start reviewing source and analyze to provide changes, but I will need that you help me with the cross compiler that you are using. I have not been able to create it.

Sure, here are the steps

1. install conan2 (i.e. via `pip install conan`)
2. download and build [mxe](https://mxe.cc/)
3. make sure in it's `settings.mk` you have 

```
MXE_TARGETS := i686-w64-mingw32.static
MXE_PLUGIN_DIRS=plugins/windows-xp
```
4. make sure that everything works, e.g. something like `make lua`, and copy lua.exe to your win-xp host and check that it launches.
5. add to your $PATH path to mxe/bin, i.e.  `i686-w64-mingw32.static-gcc` should be found
6.  created `xp` conan profile at `~/.conan2/profiles/xp`
[xp.txt](https://github.com/basiliscos/syncspirit/files/15067781/xp.txt)
7. now it should be able to pull & build all dependencies 

```
cd syncspirit
mkdir build.xp
cd build.xp
conan install --build=missing -o '*:shared=False' -o shared=False --output-folder . -s build_type=Release --profile:build=default --profile:host=xp ..
```

8. now it should `sync-spirit` can be configured & build

```
source ./conanbuild.sh
cmake .. -G "Unix Makefiles" -DCMAKE_TOOLCHAIN_FILE=$PWD/conan_toolchain.cmake -DCMAKE_POLICY_DEFAULT_CMP0091=NEW -DCMAKE_BUILD_TYPE=Release  -DCMAKE_CXX_FLAGS="-D_WIN32_WINNT=0x0501"
make -j`nproc`
```

PS. I've created `wixp-support` branch with yet not committed changes to let it build.

**Comment by basiliscos** [comment]
>  Question: What Linux version do you use for building the WinXP version?

it does not matter, any modern should be ok. I use void-linux :)

**Comment by basiliscos** [comment]
Which command did you issued and in which dir? It seems you are trying just to do `make` for the whole `mxe`...

**Comment by basiliscos** [comment]
`make lua` should be enough.

**Comment by basiliscos** [comment]
I've made a `v.0.3.1` release. Basically it is the same as the attached exe here. Also building instructions have been updated (basically the same as here). 

WBR

**Comment by basiliscos** [comment]
> [E/2868] ssdp, udp socket cancellation : The attempted operation is not supported for the type of object referenced

`BOOST_ASIO_ENABLE_CANCELIO=1` that thing does the job right :)

**Comment by basiliscos** [comment]
Hi,

I think there was a bug that dir scanning wasn't triggered. I fixed that in v0.3.2, please try. Also I specified path at `Z:\\test`, you can also try `z:\test` and other variants with normal slash.

**Comment by basiliscos** [comment]
Oh, it was windows-specific bug in my code, as it parsed `:` as parameters separator. Now it should be fixed, please try `v0.3.3`

**Comment by basiliscos** [comment]
this is expected behavior: when you delete a file on one device, and modify it on another, and then try to sync them, the 1st device does not knows what to do. It is a conflict, and it have to be resolved manually.

The same conflict situation is on syncthing. There are a few strategies to handle them: either push-only (i.e. ignore **remote** deletes/changes) changes or pull-only changes (i.e. ignore **local** deletes/changes), or make a conflict file for manual conflict resolution (default)

Neither of them is currently implemented in syncspirit, this is for future development.

**Comment by basiliscos** [comment]
this is expected behavior: when you delete a file on one device, and modify it on another, and then try to sync them, the 1st device does not knows what to do. It is a conflict, and it have to be resolved manually.

The same conflict situation is on syncthing. There are a few strategies to handle them: either push-only (i.e. ignore **remote** deletes/changes) changes or pull-only changes (i.e. ignore **local** deletes/changes), or make a conflict file for manual conflict resolution (default)

Neither of them is currently implemented in syncspirit, this is for future development.
